core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,155 @@
+		return;
+	}
+
+	finfo = (struct __ftrace_info *)(val + load_addr);
+	if (finfo->buf_end > finfo->buf_start)
+		fbuf_size = finfo->buf_end - finfo->buf_start;
+
+	if (fbuf_size >= MIN_FTRACE_BUF_SIZE) {
+		fbuf = (struct ftrace_buf *)finfo->buf_start;
+		fbuf->head_off = sizeof(struct ftrace_buf);
+		count = snprintk((char *)fbuf + fbuf->head_off,
+				 MAX_HEADER_STRLEN,
+				 "Function graph for TA: %pUl @ %lx\n",
+				 (void *)&s->ctx->uuid, load_addr);
+		assert(count < MAX_HEADER_STRLEN);
+
+		fbuf->ret_func_ptr = finfo->ret_ptr;
+		fbuf->ret_idx = 0;
+		fbuf->lr_idx = 0;
+		fbuf->buf_off = fbuf->head_off + count;
+		fbuf->curr_size = 0;
+		fbuf->max_size = fbuf_size - sizeof(struct ftrace_buf) - count;
+
+		s->fbuf_paddr = virt_to_phys(fbuf);
+		s->fbuf_size = fbuf_size;
+	} else {
+		DMSG("ftrace buffer too small");
+
+		s->fbuf_paddr = 0;
+		s->fbuf_size = 0;
+	}
+}
+
+void ta_fbuf_dump(struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf = NULL;
+	struct mobj *mobj = NULL;
+	char *va = NULL;
+	uint32_t ret = 0;
+	uint32_t dump_size = 0;
+	struct thread_param params[3] = { };
+
+	if (!s->fbuf_size)
+		return;
+
+	if (!core_mmu_add_mapping(MEM_AREA_TA_RAM, s->fbuf_paddr,
+				  s->fbuf_size)) {
+		EMSG("Failed to map TA ftrace buffer, pa: %lx, size: %lx",
+			s->fbuf_paddr, s->fbuf_size);
+		return;
+	}
+
+	fbuf = phys_to_virt(s->fbuf_paddr, MEM_AREA_TA_RAM);
+	if (!fbuf || (fbuf->curr_size > fbuf->max_size)) {
+		EMSG("Ftrace initial buffer check failed");
+		return;
+	}
+
+	if (!fbuf->curr_size)
+		DMSG("Ftrace buffer empty (no TA file compiled with -pg?)");
+
+	dump_size = fbuf->buf_off - fbuf->head_off + fbuf->curr_size;
+
+	mobj = thread_rpc_alloc_payload(sizeof(TEE_UUID) + dump_size);
+	if (!mobj) {
+		EMSG("Ftrace thread_rpc_alloc_payload failed");
+		return;
+	}
+
+	va = mobj_get_va(mobj, 0);
+	if (!va)
+		goto exit;
+
+	memcpy(va, &s->ctx->uuid, sizeof(TEE_UUID));
+	memcpy(va + sizeof(TEE_UUID), (char *)fbuf + fbuf->head_off, dump_size);
+
+	params[0] = THREAD_PARAM_VALUE(INOUT, 0, 0, 0);
+	params[1] = THREAD_PARAM_MEMREF(IN, mobj, 0, sizeof(TEE_UUID));
+	params[2] = THREAD_PARAM_MEMREF(IN, mobj, sizeof(TEE_UUID), dump_size);
+
+	ret = thread_rpc_cmd(OPTEE_RPC_CMD_FTRACE, 3, params);
+	if (ret)
+		EMSG("Ftrace thread_rpc_cmd ret: %08x", ret);
+
+exit:
+	thread_rpc_free_payload(mobj);
+}
+
+void ftrace_ta_map_lr(uint64_t *lr)
+{
+	struct ftrace_buf *fbuf = NULL;
+	struct tee_ta_session *s = NULL;
+
+	if (tee_ta_get_current_session(&s) != TEE_SUCCESS)
+		panic();
+
+	if (!s->fbuf_size)
+		return;
+
+	if (!core_mmu_add_mapping(MEM_AREA_TA_RAM, s->fbuf_paddr,
+				  s->fbuf_size)) {
+		EMSG("Failed to map TA ftrace buffer, pa: %lx, size: %lx",
+			s->fbuf_paddr, s->fbuf_size);
+		return;
+	}
+
+	fbuf = phys_to_virt(s->fbuf_paddr, MEM_AREA_TA_RAM);
+	if (!fbuf) {
+		EMSG("Failed to get ftrace buffer virtual addr");
+		return;
+	}
+
+	if ((*lr == fbuf->ret_func_ptr) && (fbuf->lr_idx < fbuf->ret_idx)) {
```

Comment:
What are we testing for and what's happening here?

Reply:
Have a look at discussion started from here: https://github.com/OP-TEE/optee_os/pull/2912#issuecomment-490808358

Reply:
Please add a comment describing what's going on.

Reply:
ok